src/AudioToolbox/MusicSequence.cs

```diff
@@ -274,6 +288,39 @@ public double GetBeatsForSeconds (double seconds)
 				return beats;
 			return 0;
 		}
+
+		[DllImport (Constants.AudioToolboxLibrary)]
+		extern static /* OSStatus */ MusicPlayerStatus MusicSequenceSetUserCallback (/* MusicSequence */ IntPtr inSequence, MusicSequenceUserCallbackProxy inCallback, /* void * */ IntPtr inClientData);
+
+		public void SetUserCallback (MusicSequenceUserCallback callback)
+		{
+			lock (userCallbackHandles) {
+				if (userCallbackHandles.ContainsKey (handle))
+					userCallbackHandles[handle] = callback;
+				else
+					userCallbackHandles.Add (handle, callback);
+			}
+
+			MusicSequenceSetUserCallback (handle, userCallbackProxy, IntPtr.Zero);
+		}
+
+#if !MONOMAC
+		[MonoPInvokeCallback (typeof (MusicSequenceUserCallbackProxy))]
+#endif
+		static void UserCallbackProxy (IntPtr inClientData, IntPtr inSequence, IntPtr inTrack, double inEventTime, IntPtr inEventData, double inStartSliceBeat, double inEndSliceBeat)
+		{
+			MusicSequenceUserCallback userCallback;
+			lock (userCallbackHandles)
+				userCallbackHandles.TryGetValue (inSequence, out userCallback);
+
+			if (userCallback != null) {
+				var userEventData = new MusicEventUserData (inEventData);
+				var musicSequence = MusicSequence.Lookup (inSequence);
+				var musicTrack = new MusicTrack (musicSequence, inTrack, false);
+
+				userCallback (musicTrack, inEventTime, userEventData, inStartSliceBeat, inEndSliceBeat);
```

Comment:
Yes, the contract is unclear, even on the native side :-( so even undisposed it's not clear if it's correct to hang on the values (and that API not refcounted afaik). The only clear thing is that we're allocating memory for some things that might never be used (by the callback) and that it will have some, *unknown* effect.

Instead of having a callback with several parameters, some costly and some likely not used, we could expose a different API with a single parameter/type that would allow getting the `MusicTrack` / `MusicEventUserData` from properties (or methods) and lazily allocate memory only when needed... that makes the creation (and if needed the dispose) part of the callback code.

@olegoid we really need a sample for this, minimally showing the rate of calls. Im fairly sure `MusicSequenceSetUserCallback` was not implemented originally (along with the rest of the API) because the "how it's being used" was unanswered.